src/pharmpy/model/external/nonmem/update.py

```diff
@@ -393,7 +393,69 @@ def update_ode_system(model: Model, old: Optional[CompartmentalSystem], new: Com
         if not is_nonlinear_odes(model):
             model = from_des(model, advan)
 
-    model, updated_dataset = update_infusion(model, old)
+    if not updated_dataset:
+        model, updated_dataset = update_infusion(model, old)
+    else:
+        model, _ = update_infusion(model, old)
+    return model, updated_dataset
+
+
+def update_cmt_column(model, old, new):
+    if model.dataset is not None:
+        if (
+            "admid" in model.datainfo.types
+            and len(model.dataset[model.datainfo.typeix["admid"].names[0]].unique()) != 1
+        ):
+            cs = model.statements.ode_system
+            newmap = new_compartmental_map(cs)
+
+            d = {}
+            for dose_comp in model.statements.ode_system.dosing_compartment:
+                d[dose_comp.dose.admid] = newmap[dose_comp.name]
+
+            from pharmpy.model.model import update_datainfo
+            from pharmpy.modeling.data import get_admid
+
+            cmt_col = get_admid(model)
+            cmt_col = cmt_col.replace(d)
+
+            dataset = model.dataset.copy()
+            dataset['CMT'] = cmt_col
+            di = update_datainfo(model.datainfo, dataset)
+            colinfo = di['CMT'].replace(type='compartment')
+            model = model.replace(datainfo=di.set_column(colinfo), dataset=dataset)
+
+            updated_dataset = True
+        elif "CMT" in model.datainfo.names and len(old.compartment_names) != len(
+            new.compartment_names
+        ):
+            # Make sure column is a number and not string
+            model.dataset["CMT"] = pd.to_numeric(model.dataset["CMT"])
+
+            # Differ in amount of compartment -> Change cmt numbering
+            # The cmt number should be the same as the dosing compartment
+            oldmap = model.internals.compartment_map
+            assert oldmap is not None
+            cs = model.statements.ode_system
+            newmap = new_compartmental_map(cs)
+            oldmap = oldmap.copy()
+            remap = create_compartment_remap(oldmap, newmap)
+
+            for dose_comp in old.dosing_compartment:
```

Comment:
As we want to remap the old doses to the new compartments we need to loop over the old ODE in order to know which dose should be moved to the corresponding new dose compartment. Naturally, it could be done the other way around, looping over the new doses and then checking which is the corresponding old dose but I found this implementation to be easier.